pyro/contrib/funsor/named_messenger.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import OrderedDict, namedtuple
+from contextlib import ExitStack  # python 3
+
+from pyro.poutine.reentrant_messenger import ReentrantMessenger
+
+
+# name_to_dim : dict, dim_to_name : dict
+StackFrame = namedtuple('StackFrame', ['name_to_dim', 'dim_to_name'])
+
+
+class NamedMessenger(ReentrantMessenger):
+    """
+    Handler for converting to/from funsors consistent with Pyro's positional batch dimensions.
+
+    :param int history: The number of previous contexts visible from the
+        current context. Defaults to 1. If zero, this is similar to
+        :class:`pyro.plate`.
+    :param bool keep: If true, frames are replayable. This is important
+        when branching: if ``keep=True``, neighboring branches at the same
+        level can depend on each other; if ``keep=False``, neighboring branches
+        are independent (conditioned on their shared ancestors).
+    """
+    def __init__(self, history=1, keep=False):
+        assert history >= 0
+        self.history = history
+        self.keep = keep
+        self._iterable = None
+        self._pos = -1
+        self._stack = []
+        super().__init__()
+
+    def generator(self, iterable):
+        self._iterable = iterable
+        return self
+
+    def __iter__(self):
+        with ExitStack() as stack:
+            for value in self._iterable:
+                stack.enter_context(self)
+                yield value
+
+    def __enter__(self):
+        self._pos += 1
+        if len(self._stack) <= self._pos:
+            self._stack.append(
+                StackFrame(name_to_dim=OrderedDict(), dim_to_name=OrderedDict()))
+        return super().__enter__()
```

Comment:
I've borrowed and modified this ministack design from `pyro.markov`, but I believe the way `NamedMessenger` currently works, with each stack frame having an environment, makes a stronger assumption about ordering of `NamedMessenger`s in the effect handler stack than does `pyro.markov`.  For example, I'm not sure this design would produce the right name<->dim mapping in the case of mutual recursion.